File: tools/whip-parity/whip-physics.js
'use strict';
// ══════════════════════════════════════════════════════════════════════════════
//  Whip physics — faithful CommonJS extraction of /tmp/OpenWhip/overlay.html
//
//  This module reproduces the Verlet rope/whip simulation EXACTLY as the JS in
//  overlay.html, with one deliberate change: every `Date.now()` is replaced by
//  an injected `now` parameter. `whipSpawnTime` is set at spawn time (the caller
//  passes it, conventionally 0). The crack detection inside `update(now)` uses
//  the injected `now`.
//
//  Only PHYSICS is ported here. Rendering helpers (catmullPoint,
//  whipSegmentBezier, draw) are intentionally excluded.
//
//  Operation ORDER is byte-for-byte faithful to the original `update()`:
//    updateHandleAim → Verlet integrate → pin handle → capSegmentStretch →
//    applyWallCollisions → applyBasePose → [constraintIters × (distance
//    constraints → applyBendLimits → applyBasePose → capSegmentStretch →
//    applyWallCollisions)] → crack detection → prevMouse = mouse.
// ══════════════════════════════════════════════════════════════════════════════

// PHYSICS-relevant constants copied verbatim from the P object in overlay.html.
// Purely-visual fields (lineWidth*, outlineWidth, handleExtraWidth,
// handleThickSegments, bgAlpha) are intentionally omitted.
const P = {
  // Rope structure
  segments:       28,     // number of chain links
  segmentLength:  25,     // base length of each link (px)
  taper:          0.6,   // tip segment is this fraction of base length

  // Physics
  gravity:        1.2,   // normal gravity
  dropGravity:    0.95,    // gravity when dropping/despawning
  damping:        0.96,  // velocity retention per frame (1 = no loss)
  constraintIters:20,     // higher = stiffer chain
  maxStretchRatio: 1.2,  // hard cap for per-link stretch during fast whips

  // Dynamic handle aim (target angle + restoring spring, not static lock)
  baseTargetAngle: -1.12, // radians, default "up-right" resting direction
  handleAimByMouseX: 0.4, // horizontal mouse movement influence on target angle
  handleAimByMouseY: 0.2, // vertical mouse movement influence on target angle
  handleAimClamp:    2.0,  // max radians target can deviate from base angle
  handleSpring:      0.7,  // restoring force to target angle
  handleAngularDamping: 0.078, // angular velocity damping
  basePoseSegments: 2,    // how many early segments are strongly guided
  basePoseStiffStart: 0.9, // stiffness near handle
  basePoseStiffEnd:   0.8, // stiffness near end of guided region

  // Elastic bend limits by chain position (handle stiff, tip floppy)
  handleMaxBendDeg: 16,   // max angle between links near handle
  tipMaxBendDeg:   130,   // max angle between links near tip
  bendRigidityStart: 0.8, // correction strength near handle
  bendRigidityEnd:   0.12, // correction strength near tip

  // Screen-edge slap
  wallBounce:      0.42,   // velocity retained after wall hit
  wallFriction:    0.86,   // tangential damping on wall hit

  // Crack detection
  crackSpeed:     260,     // tip velocity threshold to trigger crack (tuned down from 340)
  crackCooldownMs:200,   // min ms between cracks
  firstCrackGraceMs: 350, // no crack (macro) until this long after spawn

  // Initial arc shape
  arcWidth:       260,    // how far right the arc extends from mouse
  arcHeight:      185,    // how high the arc goes above mouse
};

const clamp = (v, lo, hi) => Math.max(lo, Math.min(hi, v));
const lerp = (a, b, t) => a + (b - a) * t;
const wrapPi = a => {
  while (a > Math.PI) a -= Math.PI * 2;
  while (a < -Math.PI) a += Math.PI * 2;
  return a;
};

class WhipSim {
  constructor(W, H) {
    this.W = W;
    this.H = H;
    this.whip = null;
    this.dropping = false;
    this.lastCrackTime = 0;
    this.whipSpawnTime = 0;
    this.handleAngle = P.baseTargetAngle;
    this.handleAngVel = 0;
    this.mouseX = 0;
    this.mouseY = 0;
    this.prevMouseX = 0;
    this.prevMouseY = 0;
  }

  // ── Whip creation ───────────────────────────────────────────────────────────
  // Faithful to spawnWhip(mx,my) from the JS. Returns the points array AND sets
  // this.whip / resets dropping / lastCrackTime. whipSpawnTime is the caller's
  // responsibility (set it to 0 after spawn, as the scenario runner does), so we
  // do NOT call Date.now() here.
  spawnWhip(mx, my) {
    this.dropping = false;
    this.lastCrackTime = 0;
    // NOTE: in the JS this is `whipSpawnTime = Date.now()`. We leave whipSpawnTime
    // to the caller (conventionally 0) to keep the simulation deterministic.
    const pts = [];
    for (let i = 0; i < P.segments; i++) {
      const t = i / (P.segments - 1);
      // Nice upward arc from handle (mouse) to tip
      const x = mx + t * P.arcWidth;
      const y = my - Math.sin(t * Math.PI * 0.75) * P.arcHeight;
      pts.push({ x, y, px: x, py: y });
    }
    this.whip = pts;
    return pts;
  }

  segLen(i) {
    const t = i / (P.segments - 1);
    return P.segmentLength * (1 - t * (1 - P.taper));
  }

  updateHandleAim() {
    if (this.dropping) return;
    const mvx = this.mouseX - this.prevMouseX;
    const mvy = this.mouseY - this.prevMouseY;
    const delta = clamp(
      mvx * P.handleAimByMouseX + mvy * P.handleAimByMouseY,
      -P.handleAimClamp,
      P.handleAimClamp
    );
    const target = P.baseTargetAngle + delta;
    const err = wrapPi(target - this.handleAngle);
    this.handleAngVel += err * P.handleSpring;
    this.handleAngVel *= P.handleAngularDamping;
    this.handleAngle = wrapPi(this.handleAngle + this.handleAngVel);
  }

  applyBasePose() {
    const whip = this.whip;
    if (!whip || this.dropping) return;
    const dx = Math.cos(this.handleAngle);
    const dy = Math.sin(this.handleAngle);
    const guided = Math.min(P.basePoseSegments, whip.length - 1);
    for (let i = 1; i <= guided; i++) {
      const t = (i - 1) / Math.max(guided - 1, 1);
      const stiff = lerp(P.basePoseStiffStart, P.basePoseStiffEnd, t);
      const prev = whip[i - 1];
      const p = whip[i];
      const targetLen = this.segLen(i - 1);
      const tx = prev.x + dx * targetLen;
      const ty = prev.y + dy * targetLen;
      p.x = lerp(p.x, tx, stiff);
      p.y = lerp(p.y, ty, stiff);
    }
  }

  applyBendLimits() {
    const whip = this.whip;
    if (!whip || whip.length < 3) return;
    for (let i = 1; i < whip.length - 1; i++) {
      const a = whip[i - 1];
      const b = whip[i];
      const c = whip[i + 1];

      const v1x = a.x - b.x;
      const v1y = a.y - b.y;
      const v2x = c.x - b.x;
      const v2y = c.y - b.y;
      // sqrt(a*a+b*b) form (not Math.hypot) to bit-match the Swift port.
      const l1 = Math.sqrt(v1x * v1x + v1y * v1y) || 0.0001;
      const l2 = Math.sqrt(v2x * v2x + v2y * v2y) || 0.0001;
      const n1x = v1x / l1, n1y = v1y / l1;
      const n2x = v2x / l2, n2y = v2y / l2;

      const dot = clamp(n1x * n2x + n1y * n2y, -1, 1);
      const angle = Math.acos(dot);
      const t = i / (whip.length - 2);
      const maxBend = lerp(P.handleMaxBendDeg, P.tipMaxBendDeg, t) * Math.PI / 180;
      const bend = Math.PI - angle; // bend away from a straight line
      if (bend <= maxBend) continue;

      // Clamp to max bend while preserving side/sign of the bend.
      const cross = n1x * n2y - n1y * n2x;
      const sign = cross >= 0 ? 1 : -1;
      const targetAngle = Math.PI - maxBend;
      const targetA = Math.atan2(n1y, n1x) + sign * targetAngle;
      const tx = b.x + Math.cos(targetA) * l2;
      const ty = b.y + Math.sin(targetA) * l2;
      const rigidity = lerp(P.bendRigidityStart, P.bendRigidityEnd, t);

      c.x = lerp(c.x, tx, rigidity);
      c.y = lerp(c.y, ty, rigidity);
    }
  }

  capSegmentStretch() {
    const whip = this.whip;
    if (!whip || whip.length < 2) return;
    for (let i = 0; i < whip.length - 1; i++) {
      const a = whip[i];
      const b = whip[i + 1];
      const dx = b.x - a.x;
      const dy = b.y - a.y;
      const dist = Math.sqrt(dx * dx + dy * dy) || 0.0001;
      const maxLen = this.segLen(i) * P.maxStretchRatio;
      if (dist <= maxLen) continue;
      const k = maxLen / dist;
      b.x = a.x + dx * k;
      b.y = a.y + dy * k;
    }
  }

  applyWallCollisions() {
    const whip = this.whip;
    if (!whip || this.dropping) return; // disable collisions while dropping
    const W = this.W;
    const H = this.H;
    const start = 1; // keep pinned handle untouched
    for (let i = start; i < whip.length; i++) {
      const p = whip[i];
      let vx = p.x - p.px;
      let vy = p.y - p.py;
      let hit = false;

      if (p.x < 0) {
        p.x = 0;
        if (vx < 0) vx = -vx * P.wallBounce;
        vy *= P.wallFriction;
        hit = true;
      } else if (p.x > W) {
        p.x = W;
        if (vx > 0) vx = -vx * P.wallBounce;
        vy *= P.wallFriction;
        hit = true;
      }

      if (p.y < 0) {
        p.y = 0;
        if (vy < 0) vy = -vy * P.wallBounce;
        vx *= P.wallFriction;
        hit = true;
      } else if (p.y > H) {
        p.y = H;
        if (vy > 0) vy = -vy * P.wallBounce;
        vx *= P.wallFriction;
        hit = true;
      }

      if (hit) {
        p.px = p.x - vx;
        p.py = p.y - vy;
      }
    }
  }

  // ── Physics step ────────────────────────────────────────────────────────────
  // Faithful port of update(). `now` (ms) is injected and used for crack
  // detection. Returns true iff a crack fired this call.
  update(now) {
    const whip = this.whip;
    if (!whip) return false;

    let crackFired = false;

    const g = this.dropping ? P.dropGravity : P.gravity;
    this.updateHandleAim();

    // Verlet integration
    const start = this.dropping ? 0 : 1; // if dropping, handle is free too
    for (let i = start; i < whip.length; i++) {
      const p = whip[i];
      const vx = (p.x - p.px) * P.damping;
      const vy = (p.y - p.py) * P.damping;
      p.px = p.x;
      p.py = p.y;
      p.x += vx;
      p.y += vy + g;
    }

    // Pin handle to mouse
    if (!this.dropping) {
      whip[0].x = this.mouseX;
      whip[0].y = this.mouseY;
      whip[0].px = this.mouseX;
      whip[0].py = this.mouseY;
    }

    // Prevent rubber-band stretching spikes before constraints.
    this.capSegmentStretch();
    this.applyWallCollisions();

    // Keep early whip segments posed upward from handle.
    this.applyBasePose();

    // Distance constraints (multiple iterations for stiffness)
    for (let iter = 0; iter < P.constraintIters; iter++) {
      for (let i = 0; i < whip.length - 1; i++) {
        const a = whip[i], b = whip[i + 1];
        const dx = b.x - a.x, dy = b.y - a.y;
        const dist = Math.sqrt(dx * dx + dy * dy) || 0.0001;
        const target = this.segLen(i);
        const diff = (dist - target) / dist * 0.5;
        const ox = dx * diff, oy = dy * diff;
        if (i === 0 && !this.dropping) {
          // Handle is pinned – push only the next point
          b.x -= ox * 2;
          b.y -= oy * 2;
        } else {
          a.x += ox; a.y += oy;
          b.x -= ox; b.y -= oy;
        }
      }
      // Clamp bend angle per joint; near handle = stiffer, near tip = floppier.
      this.applyBendLimits();
      if (!this.dropping) this.applyBasePose();
      this.capSegmentStretch();
      this.applyWallCollisions();
    }

    // Tip velocity for crack detection. sqrt(a*a+b*b) form to bit-match Swift.
    const tip = whip[whip.length - 1];
    const tdx = tip.x - tip.px;
    const tdy = tip.y - tip.py;
    const tipVel = Math.sqrt(tdx * tdx + tdy * tdy);

    if (!this.dropping && tipVel > P.crackSpeed) {
      // `now` is injected (was Date.now() in the JS).
      if (now - this.whipSpawnTime >= P.firstCrackGraceMs && now - this.lastCrackTime > P.crackCooldownMs) {
        this.lastCrackTime = now;
        // playCrackSound() + window.bridge.whipCrack() are side effects we model
        // as "a crack fired this frame".
        crackFired = true;
      }
    }

    // If dropping, check if everything fell off screen
    if (this.dropping && whip.every(p => p.y > this.H + 60)) {
      this.whip = null;
      this.dropping = false;
    }
    this.prevMouseX = this.mouseX;
    this.prevMouseY = this.mouseY;

    return crackFired;
  }
}

module.exports = { WhipSim, P, clamp, lerp, wrapPi };

File: Sources/VictorAddons/WhipPhysics.swift
import Foundation

/// Pure, deterministic Verlet whip physics — a faithful Swift port of the
/// JavaScript simulation in OpenWhip's `overlay.html` (https://github.com/GitFrog1111/OpenWhip).
///
/// This type holds NO UI or platform dependency so it can be unit-tested and
/// compared frame-by-frame against the original JS via golden values
/// (see Tests/VictorAddonsTests/WhipPhysicsTests + tools/whip-parity).
///
/// Coordinate space matches the original HTML canvas: origin top-left, +y points
/// DOWN, so gravity is +y. The rendering view is flipped to match.
///
/// The operation ORDER inside `update(now:)` is kept byte-for-byte with the JS
/// (same loop bounds, same sequence of stretch/wall/base-pose calls inside and
/// outside the constraint-iteration loop). Changing the order breaks parity.
final class WhipPhysics {

    /// Physics constants — copied verbatim from the `P` object in overlay.html
    /// (physics-relevant fields only; visual fields like line widths live in the view).
    enum P {
        // Rope structure
        static let segments = 28          // number of chain links
        static let segmentLength = 25.0   // base length of each link (px)
        static let taper = 0.6            // tip segment is this fraction of base length

        // Physics
        static let gravity = 1.2
        static let dropGravity = 0.95
        static let damping = 0.96
        static let constraintIters = 20
        static let maxStretchRatio = 1.2

        // Dynamic handle aim
        static let baseTargetAngle = -1.12
        static let handleAimByMouseX = 0.4
        static let handleAimByMouseY = 0.2
        static let handleAimClamp = 2.0
        static let handleSpring = 0.7
        static let handleAngularDamping = 0.078
        static let basePoseSegments = 2
        static let basePoseStiffStart = 0.9
        static let basePoseStiffEnd = 0.8

        // Elastic bend limits by chain position
        static let handleMaxBendDeg = 16.0
        static let tipMaxBendDeg = 130.0
        static let bendRigidityStart = 0.8
        static let bendRigidityEnd = 0.12

        // Screen-edge slap
        static let wallBounce = 0.42
        static let wallFriction = 0.86

        // Crack detection
        static let crackSpeed = 260.0     // tip velocity threshold to trigger a crack
                                          // (tuned down from OpenWhip's 340 so a
                                          //  gentler flick still cracks)
        static let crackCooldownMs = 200.0
        static let firstCrackGraceMs = 350.0

        // Programmatic-crack impulse (Enter-button crack; not part of the JS port)
        static let forceCrackKickX = 80.0   // forward (+x) tip velocity, px/frame
        static let forceCrackKickY = 160.0  // upward (-y) tip velocity, px/frame

        // Initial arc shape
        static let arcWidth = 260.0
        static let arcHeight = 185.0
    }

    /// One chain link. `(px,py)` is the previous position used by Verlet integration.
    struct Point: Equatable {
        var x: Double
        var y: Double
        var px: Double
        var py: Double
    }

    // MARK: - State (all mirrors the JS globals)

    private(set) var points: [Point] = []
    private(set) var isActive = false
    private(set) var dropping = false

    private var lastCrackTime = 0.0
    private var whipSpawnTime = 0.0
    private var handleAngle = P.baseTargetAngle
    private var handleAngVel = 0.0

    private(set) var mouseX = 0.0
    private(set) var mouseY = 0.0
    private var prevMouseX = 0.0
    private var prevMouseY = 0.0

    var W: Double
    var H: Double

    init(width: Double, height: Double) {
        self.W = width
        self.H = height
    }

    // MARK: - Lifecycle

    /// Spawn the whip with the handle at the cursor — mirrors `spawnWhip` + `onSpawnWhip`.
    func spawn(mouseX mx: Double, mouseY my: Double, now: Double) {
        mouseX = mx
        mouseY = my
        prevMouseX = mx
        prevMouseY = my
        dropping = false
        lastCrackTime = 0
        whipSpawnTime = now
        handleAngle = P.baseTargetAngle
        handleAngVel = 0

        var pts: [Point] = []
        pts.reserveCapacity(P.segments)
        for i in 0..<P.segments {
            let t = Double(i) / Double(P.segments - 1)
            let x = mx + t * P.arcWidth
            let y = my - sin(t * Double.pi * 0.75) * P.arcHeight
            pts.append(Point(x: x, y: y, px: x, py: y))
        }
        points = pts
        isActive = true
    }

    /// Begin the drop/despawn animation — mirrors `onDropWhip`.
    func startDropping() {
        if isActive && !dropping { dropping = true }
    }

    func setMouse(_ x: Double, _ y: Double) {
        mouseX = x
        mouseY = y
    }

    /// Programmatically crack the whip with NO mouse motion (driven by the
    /// Enter-button while the overlay is up). Injects a whip-snap velocity into
    /// the free chain — concentrated toward the tip, thrown up-and-forward — so
    /// it visibly flicks, and arms the crack cooldown so the induced motion
    /// doesn't also trip the natural detector. Not part of the JS port.
    func crackImpulse(now: Double) {
        guard isActive, !dropping, points.count >= 4 else { return }
        lastCrackTime = now
        let n = points.count
        let base = P.basePoseSegments          // leave the handle/base pose intact
        let span = Double(n - 1 - base)
        guard span > 0 else { return }
        for i in base..<n {
            let t = Double(i - base) / span     // 0 at base → 1 at tip
            let s = t * t                       // concentrate the snap at the tip
            points[i].px -= P.forceCrackKickX * s   // +x velocity (forward)
            points[i].py += P.forceCrackKickY * s   // -y velocity (up)
        }
    }

    // MARK: - Per-link length / helpers

    private func segLen(_ i: Int) -> Double {
        let t = Double(i) / Double(P.segments - 1)
        return P.segmentLength * (1 - t * (1 - P.taper))
    }

    private func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double {
        Swift.max(lo, Swift.min(hi, v))
    }

    private func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
        a + (b - a) * t
    }

    private func wrapPi(_ angle: Double) -> Double {
        var a = angle
        while a > Double.pi { a -= Double.pi * 2 }
        while a < -Double.pi { a += Double.pi * 2 }
        return a
    }

    // MARK: - Physics step (mirrors `update()`)

    /// Advance one frame. Returns `true` if a crack fired this frame (tip velocity
    /// exceeded the threshold, past the spawn grace and crack cooldown). `now` is
    /// milliseconds (injected so the simulation is deterministic and testable).
    @discardableResult
    func update(now: Double) -> Bool {
        guard isActive else { return false }

        let g = dropping ? P.dropGravity : P.gravity
        updateHandleAim()

        // Verlet integration
        let start = dropping ? 0 : 1
        if start < points.count {
            for i in start..<points.count {
                let vx = (points[i].x - points[i].px) * P.damping
                let vy = (points[i].y - points[i].py) * P.damping
                points[i].px = points[i].x
                points[i].py = points[i].y
                points[i].x += vx
                points[i].y += vy + g
            }
        }

        // Pin handle to mouse
        if !dropping {
            points[0].x = mouseX
            points[0].y = mouseY
            points[0].px = mouseX
            points[0].py = mouseY
        }

        capSegmentStretch()
        applyWallCollisions()
        applyBasePose()

        // Distance constraints (multiple iterations for stiffness)
        for _ in 0..<P.constraintIters {
            for i in 0..<(points.count - 1) {
                let dx = points[i + 1].x - points[i].x
                let dy = points[i + 1].y - points[i].y
                var dist = (dx * dx + dy * dy).squareRoot()
                if dist == 0 { dist = 0.0001 }
                let target = segLen(i)
                let diff = (dist - target) / dist * 0.5
                let ox = dx * diff
                let oy = dy * diff
                if i == 0 && !dropping {
                    points[i + 1].x -= ox * 2
                    points[i + 1].y -= oy * 2
                } else {
                    points[i].x += ox
                    points[i].y += oy
                    points[i + 1].x -= ox
                    points[i + 1].y -= oy
                }
            }
            applyBendLimits()
            if !dropping { applyBasePose() }
            capSegmentStretch()
            applyWallCollisions()
        }

        // Tip velocity for crack detection. Use the sqrt(a*a+b*b) form (not
        // hypot/pow) to bit-match the JS golden harness.
        let tip = points[points.count - 1]
        let tdx = tip.x - tip.px
        let tdy = tip.y - tip.py
        let tipVel = (tdx * tdx + tdy * tdy).squareRoot()

        var crack = false
        if !dropping && tipVel > P.crackSpeed {
            if now - whipSpawnTime >= P.firstCrackGraceMs && now - lastCrackTime > P.crackCooldownMs {
                lastCrackTime = now
                crack = true
            }
        }

        // If dropping, check if everything fell off screen
        if dropping && points.allSatisfy({ $0.y > H + 60 }) {
            points = []
            isActive = false
            dropping = false
        }

        prevMouseX = mouseX
        prevMouseY = mouseY
        return crack
    }

    // MARK: - Sub-steps (each mirrors the same-named JS function)

    private func updateHandleAim() {
        if dropping { return }
        let mvx = mouseX - prevMouseX
        let mvy = mouseY - prevMouseY
        let delta = clamp(
            mvx * P.handleAimByMouseX + mvy * P.handleAimByMouseY,
            -P.handleAimClamp,
            P.handleAimClamp
        )
        let target = P.baseTargetAngle + delta
        let err = wrapPi(target - handleAngle)
        handleAngVel += err * P.handleSpring
        handleAngVel *= P.handleAngularDamping
        handleAngle = wrapPi(handleAngle + handleAngVel)
    }

    private func applyBasePose() {
        guard isActive, !dropping else { return }
        let dx = cos(handleAngle)
        let dy = sin(handleAngle)
        let guided = Swift.min(P.basePoseSegments, points.count - 1)
        guard guided >= 1 else { return }
        for i in 1...guided {
            let t = Double(i - 1) / Double(Swift.max(guided - 1, 1))
            let stiff = lerp(P.basePoseStiffStart, P.basePoseStiffEnd, t)
            let prevX = points[i - 1].x
            let prevY = points[i - 1].y
            let targetLen = segLen(i - 1)
            let tx = prevX + dx * targetLen
            let ty = prevY + dy * targetLen
            points[i].x = lerp(points[i].x, tx, stiff)
            points[i].y = lerp(points[i].y, ty, stiff)
        }
    }

    private func applyBendLimits() {
        guard points.count >= 3 else { return }
        for i in 1..<(points.count - 1) {
            let ax = points[i - 1].x, ay = points[i - 1].y
            let bx = points[i].x, by = points[i].y
            let cx = points[i + 1].x, cy = points[i + 1].y

            let v1x = ax - bx
            let v1y = ay - by
            let v2x = cx - bx
            let v2y = cy - by
            var l1 = (v1x * v1x + v1y * v1y).squareRoot(); if l1 == 0 { l1 = 0.0001 }
            var l2 = (v2x * v2x + v2y * v2y).squareRoot(); if l2 == 0 { l2 = 0.0001 }
            let n1x = v1x / l1, n1y = v1y / l1
            let n2x = v2x / l2, n2y = v2y / l2

            let dot = clamp(n1x * n2x + n1y * n2y, -1, 1)
            let angle = acos(dot)
            let t = Double(i) / Double(points.count - 2)
            let maxBend = lerp(P.handleMaxBendDeg, P.tipMaxBendDeg, t) * Double.pi / 180
            let bend = Double.pi - angle
            if bend <= maxBend { continue }

            let cross = n1x * n2y - n1y * n2x
            let sign: Double = cross >= 0 ? 1 : -1
            let targetAngle = Double.pi - maxBend
            let targetA = atan2(n1y, n1x) + sign * targetAngle
            let tx = bx + cos(targetA) * l2
            let ty = by + sin(targetA) * l2
            let rigidity = lerp(P.bendRigidityStart, P.bendRigidityEnd, t)

            points[i + 1].x = lerp(cx, tx, rigidity)
            points[i + 1].y = lerp(cy, ty, rigidity)
        }
    }

    private func capSegmentStretch() {
        guard points.count >= 2 else { return }
        for i in 0..<(points.count - 1) {
            let dx = points[i + 1].x - points[i].x
            let dy = points[i + 1].y - points[i].y
            var dist = (dx * dx + dy * dy).squareRoot()
            if dist == 0 { dist = 0.0001 }
            let maxLen = segLen(i) * P.maxStretchRatio
            if dist <= maxLen { continue }
            let k = maxLen / dist
            points[i + 1].x = points[i].x + dx * k
            points[i + 1].y = points[i].y + dy * k
        }
    }

    private func applyWallCollisions() {
        guard isActive, !dropping else { return } // disable collisions while dropping
        let start = 1 // keep pinned handle untouched
        for i in start..<points.count {
            var vx = points[i].x - points[i].px
            var vy = points[i].y - points[i].py
            var hit = false

            if points[i].x < 0 {
                points[i].x = 0
                if vx < 0 { vx = -vx * P.wallBounce }
                vy *= P.wallFriction
                hit = true
            } else if points[i].x > W {
                points[i].x = W
                if vx > 0 { vx = -vx * P.wallBounce }
                vy *= P.wallFriction
                hit = true
            }

            if points[i].y < 0 {
                points[i].y = 0
                if vy < 0 { vy = -vy * P.wallBounce }
                vx *= P.wallFriction
                hit = true
            } else if points[i].y > H {
                points[i].y = H
                if vy > 0 { vy = -vy * P.wallBounce }
                vx *= P.wallFriction
                hit = true
            }

            if hit {
                points[i].px = points[i].x - vx
                points[i].py = points[i].y - vy
            }
        }
    }
}
